piptools/repositories/pypi.py

```diff
@@ -157,15 +144,16 @@ def find_best_match(self, ireq, prereleases=None):
         if not matching_candidates:
             raise NoCandidateFound(ireq, all_candidates, self.finder)
 
-        # pip <= 19.0.3
-        if hasattr(self.finder, "_candidate_sort_key"):
+        if get_pip_version() < parse_version("19.1"):
             best_candidate = max(
                 matching_candidates, key=self.finder._candidate_sort_key
             )
-        # pip >= 19.1
-        else:
+        elif get_pip_version() < parse_version("19.2"):
```

Comment:
I'd cache `get_pip_version()` here (i.e. only call it once).
Or even have it somewhere at the module level already (assuming it gets used almost always).

This could also be stored/used similar to PY2/PY3 often is, e.g. `IS_PIP_19_2` maybe.

Reply:
If add to `piptools/utils.py`

```python
PIP_VERSION = parse_version(pip.__version__):
```

and compare by:

```python
from piptools.utils import PIP_VERSION

if PIP_VERSION < parse_version("19.1"):
   ...
```

What do you think?

Reply:
Or convert PIP_VERSION to a tuple like `(19, 2, 1)` and compare by:

```python
if PIP_VERSION < (19, 1):
   ...
```

Reply:
I like the tuple idea.

Reply:
For performance reasons you might want to peak at https://github.com/pypa/pip/blob/d8c992acc3c8477488bd3b9aa4b13468d1d0b421/src/pip/__init__.py#L1 directly, falling back if it fails maybe.
(but have not checked how much/often it is used actually)

Reply:
Fixed to tuple comparing.